src/main/kotlin/pl/edu/agh/server/domain/authentication/AuthenticationService.kt

```diff
@@ -97,10 +110,52 @@ class AuthenticationService(
         tokenRepository.save(refreshToken.get())
     }
 
+    @Transactional
     fun logout(refreshToken: String) {
         val user = userRepository.findByEmail(jwtService.extractUsername(refreshToken))
             .orElseThrow { throw Exception("User not found") }
         revokeAllUserAccessTokens(user.id!!)
         revokeRefreshToken(refreshToken)
     }
+
+    fun verifyEmail(verificationToken: String) {
+        if (jwtService.isVerificationTokenValid(verificationToken).not()) {
+            throw IllegalArgumentException("Invalid verification token")
+        }
+
+        val user = userRepository.findByVerificationToken(verificationToken)
+            .orElseThrow { throw IllegalArgumentException("Invalid verification token") }
+
+        user.enabled = true
+        user.verificationToken = null
+        userRepository.save(user)
+    }
+
+    @Transactional
+    fun prepareForPasswordChange(request: ResetPasswordRequest) {
+        val user =
+            userRepository.findByEmail(request.email).orElseThrow { throw IllegalArgumentException("Invalid email") }
```

Comment:
To wymagaloby raczej duzej zmiany i troche roboty, wiec mozesz smialo zignorowac, pisze bardziej jako ciekawostke.

Zwracanie bledu gdy uzytkownik poda niewlasciwego maila umozliwi enumeracje po mailach i ulatwi potencjalny przyszly brute force hasel. IMO powinno sie zwracac zawsze generyczna wiadomosc: "Jezeli podane dane sa prawidze to powinenes dostac maila". Dodatkowo czas odpowiedzi powinen byc nieodroznialny w przypadku poprawnego i niepoprawnego maila. W tym wypadku nawet jezeli wyjdziesz z funkcji gdy podano niewlasciwy email zamiast rzucic blad, to prawdopodobnie podanie w requescie dlugiego (200,300,400... znakow) hasla sprawi ze jego dalsze encodowanie zrobi duza roznice czasowa i dalej bedzie mozna enumerowac nazwy uzytkownikow sprawdzajac czas jaki dany request byl procesowany.

Problem enumeracji uztkownikow jest tez w innych funkcjach jak widze.

To by było moglo byc jako medium podatnosc zgloszone: https://docs.fluidattacks.com/criteria/vulnerabilities/026/